### DISCUSSION TITLE
Is there a way to generate a mind map from markdown?

### DISCUSSION THREAD

**Discussion by Francespo** [freeplane/freeplane#1821] [Help]
title

**Comment by macmarrum** [comment]
Short answer: there is – you can write a script.

Long answer: it depends on what you expect as the end result. The simmplest import is your Markdown code in a single node with `Format: Markdown`, so that Freeplane can display it as is.

**Reply by Francespo** [reply]
I'd like to set a template for the map in freeplane, then importing the markdown either with an indented bullet list or made up of headings and have freeplane generate a map with 
# Heading 0 = root node
## Heading 1 = level 1 node
## Heading 2 = level 1 node
### Heading 2.1 = level 2 node (child of Heading 2)

**Comment by macmarrum** [comment]
You can try the following:

1. Download [Import.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/io/github/macmarrum/freeplane/Import.groovy) and save it to `<user-dir>/lib/`
2. Download [Export.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/io/github/macmarrum/freeplane/Export.groovy) and save it to `<user-dir>/lib/`
3. Download [importFromMarkdownClpb.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/importFromMarkdownClpb.groovy)  and save it to `<user-dir>/scripts/`
4. Restart Freeplane
6. Copy your Markdown (code) to clipboard
7. Back in Freeplane, open a mind map (e.g. create one from a template or your choosing)
8. Select a node under which you want your Markdown to be imported (e.g. the root node)
9. Execute the script `Mac1 -> Import -> Import From Markdown Clpb`

You might want to enable `Format->Automatic level styles->for all nodes` to have your heading nodes "light up".

**Note**: *You can find your `<user-dir>` in Freeplane `Tools -> Open user directory`*

**Reply by cobblepot1** [reply]
I'm looking for an easy way to customize Freeplane to Markdown export, would these scripts provide a way for me to do that? I want to control things like Freeplane exporting node levels as tabs rather than as double-spaces.

**Reply by macmarrum** [reply]
Just to cover all the bases, have you tried the standard `File->Export map…` "Markdown" functionality?

My [Export.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/io/github/macmarrum/freeplane/Export.groovy) library contains supports several export formats, including Markdown. The procedure is automated in [exportToMarkdownFile.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/exportToMarkdownFile.groovy). It exports the branch of the selected node.

*You'll need [Export.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/io/github/macmarrum/freeplane/Export.groovy) in your `<user-dir>/lib/`*

**Reply by cobblepot1** [reply]
Thanks! I have tried Export map... to markdown, but I can't tell what file transforms the map to Markdown so I don't know how to customize it. I was hoping that your script would be easier for me to customize but in looking at it I see I'm not knowledgeable enough to do it at this point.

**Reply by macmarrum** [reply]
If you attach an example mind map (as a ZIP file) and the Markdown code that you expect when you export the mind map, maybe I'll be able to help.

**Reply by cobblepot1** [reply]
OK, attached is a map, the actual export output, what my goal output is, and what I get when I copy/paste from Freeplane to VS Code (which is actually much closer to my goal output than the 'official' export).

I would like a script that I can change to make a custom export style. I know CSS and Markdown but not Java or Groovy or anything like that.
[markdown-test.zip](https://github.com/freeplane/freeplane/files/15284702/markdown-test.zip)

**Reply by macmarrum** [reply]
I've updated [Export.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/io/github/macmarrum/freeplane/Export.groovy) to consider bullet items.

You can use non-Markdown nodes for bullets too, instead use a style (either `ulBullet` or `olBullet`).
Similarly, you can use Level styles for headings.

Here's a sample mind map 
[markdown-macmarrum.zip](https://github.com/freeplane/freeplane/files/15287831/markdown-macmarrum.zip)

**Comment by cobblepot1** [comment]
I've had success just copy/pasting from a text editor into Freeplane. Tabs import as node levels. I use hyphen bullets so I need to search/replace `- ` with [blank] to remove the bullets from the start of each node. You can probably set up auto-format by level to get the formatting you want.
